working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+should be extended with shareable versions of these and a way to convert an
+existing object to a shareable one.
+
+```dart
+class ShareableFuture<T implements Shareable?>
+    implements Future<T>, Shareable {
+}
+
+class ShareableStream<T implements Shareable?>
+    implements Stream<T>, Shareable {
+}
+
+extension ToShareableFuture<E extends Shareable?> on Future<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableFuture<E> toShareable() => /* ... */;
+}
+
+extension ToShareableStream<E extends Shareable?> on Stream<E> {
+  /// Converts the [Future] to [ShareableFuture] if it is not already shareable.
+  ShareableStream<E> toShareable() => /* ... */;
+}
+```
+
+The reason for providing these is to allow developers to structure their code
+using well understood primitives futures and streams instead of devising new
+primitives which reimplement some of the same functionality and is compatible
+with threading.
+
+#### Executors of async callbacks
+
+Consider the following code:
+
+```dart
+Future<void> foo() async {
+  await something();
+  print(1);
+}
+
+void main() async {
+  await Isolate.runShared(() async {
+    await foo();
+  });
+}
+```
+
+What happens when `Future` completes in a shared isolate? Who drives event loop
+of that isolate? Which thread will callbacks run on?
+
+I propose to introduce another concept similar to `Zone`: `Executor`. Executors
+encapsulate the notion of the event loop and control how tasks are executed.
+
+```dart
+abstract interface class Executor implements Shareable {
+    /// Returns the current executor.
+    static Executor get current;
+
+    Isolate get owner;
+
+    /// Schedules the given task to run in the given executor.
+    ///
+    /// If the current isolate is not the [owner] of this executor
+    /// the behavior depends on [copy]:   ///
+    ///   * When [copy] is `false` (which is default) [task] is
+    ///     expected to be [Shareable] and is transfered to the
+    ///     target isolate directly. If it is not [Shareable] then
+    ///     `ArgumentError` will be thrown.
+    ///   * If [copy] is `true` then [task] is copied to the
+    ///     target isolate using the same algorithm [SendPort.send]
+    ///     uses.
+    ///
+    void schedule(void Function() task,
+                  {bool copy = false});
+}
+```
+
+How a particular executor runs scheduled tasks depends on the executor itself:
+e.g. an executor can have a pool of threads or notify an embedder that it has
+tasks to run and let embedder run these tasks.
+
+All built-in asynchronous primitives will make the following API guarantee: **a
+callback passed to `Future` or `Stream` APIs will be invoked using executor
+which was running the code which registered the callback.**
+
+In other words `fut1 = fut.then(cb)` is equivalent to:
+
+```dart
+final result = ShareableBox(Completer<R>());
+final callback = ShareableBox(cb);
+final executor = Executor.current;
+fut.then((v) {
```

Comment:
So functions are shareable if they don't close over non-shareable state? Including accessing non-shared global variables?

Reply:
Correct. This is documented in the `Functions` section.